### PR TITLE
node: metrics: add resource alignment metrics

### PR DESCRIPTION
**Pull request by ffromani**
/kind feature

#### What this PR does / why we need it:
Add metrics to give more feedback about aligned allocation of compute resources (xref: https://github.com/kubernetes/enhancements/pull/4810#discussion_r1738664727 https://github.com/kubernetes/enhancements/pull/4810#discussion_r1740964156)

#### Which issue(s) this PR fixes:
Fixes N/A

#### Special notes for your reviewer:
The current dimension, meant as possible valueset of labels, is currently 2 (two). We plan to add more possible alignments (e.g. https://github.com/kubernetes/enhancements/pull/4810) in the future, but this number is expected to grow both very slowly and very rarely. I'd expect the dimension to stay well below 10 (ten) in the very worst case.

#### Does this PR introduce a user-facing change?
```release-note
Added a kubelet metric `container_aligned_compute_resources_count`  to report the count of containers getting aligned compute resources
```

### CODE REVIEW

**Comment by ffromani** [comment]
/sig node
/priority backlog

priority may be raised if turns to be useful (conversation ongoing) with the related KEP

**Comment by ffromani** [comment]
> ok, as weird as it seems to me, will check locally before more retests

it seems it was actually all CI issues

**Comment by ffromani** [comment]
> > ok, as weird as it seems to me, will check locally before more retests
> 
> it seems it was actually all CI issues

Maybe, but surely had a bug mismatching labels. Fixed in the latest upload.

**Comment by bart0sh** [comment]
/assign @mrunalp @SergeyKanzhelev 
for approval

**Comment by SergeyKanzhelev** [comment]
@ffromani can you please beef up the release note text. Something like: "Metric `container_aligned_compute_resources_count` with dimension `xx` was introduced that will be used to blah." It will make release team job easier closer to release time

**Review by SergeyKanzhelev — COMMENTED** [review]
looks good overall. I like having a metric for this. I have hard time to reason how accurate it will be based on your comment in code. Do you have this understanding?

**Comment by ffromani** [comment]
/hold

@SergeyKanzhelev thanks for the review, good points (relnote too terse and some concern about accuracy). I'll reword and clarify both so we can iterate.

**Comment by ffromani** [comment]
@SergeyKanzhelev I think I addressed both your last points in my latest upload, PTAL

**Comment by ffromani** [comment]
/hold

want to run a bit extra verification on my side

**Comment by ffromani** [comment]
/retest-required
```
{Failed;  === RUN   TestTerminationOrderingObeysGrace
    kuberuntime_termination_order_test.go:278: expected delay for container sc3 = 1, got 0
--- FAIL: TestTerminationOrderingObeysGrace (3.00s)
```
seems very much unrelated

File: pkg/kubelet/cm/topologymanager/scope_pod.go
Comment (haircommander) [inline pkg/kubelet/cm/topologymanager/scope_pod.go:67]: should we separate pod and container alignment? do we expect a container to be aligned IFF a pod is aligned? if so, will the value of 2 per aligned container be confusing?

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/scope_pod.go:67]: The semantic can be confusing but I'm struggling to make it easier to consume if one's not deep into details (which will likely bring in turn questions about the design of the features, but let's not digress now).

with scope=pod, the topology manager (TM) will place ALL the pod's containers on the same NUMA zone, failing to admit if not possible. So if we are here, it means all the containers are BOTH aligned AND all on the _same_ NUMA zone.

with scope=container, TM will place ALL the pod's containers on ANY available NUMA zone, failing to admit if any container can't get aligned resources. So, if we are here, it means ALL containers got NUMA aligned resources, but they can be scattered on different NUMA zones.

Comment (haircommander) [reply pkg/kubelet/cm/topologymanager/scope_pod.go:67]: so I guess the question is: should the metric footprint look different for those two cases?  would an admin want to quicky tell what it meant by alignedNUMAZone? if they have distinct code paths and distinct results, it seems like something that would warrant an  additional label value

Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/scope_pod.go:67]: good point. I can add a metric label "scope" whose value can be "pod" or "container"

**Comment by ffromani** [comment]
/hold

need to adjust the tests

**Comment by SergeyKanzhelev** [comment]
@ffromani please ping when it's time to review

**Comment by ffromani** [comment]
@SergeyKanzhelev ready for review!

**Comment by ffromani** [comment]
pull-kubernetes-node-kubelet-serial-crio-cgroupv1 seems to have failed for infra issues;on top of that, there's nothing cgroup specific, or even cgroup-related, in this PR.

**Comment by ffromani** [comment]
serial-crio-cgroupv* lane failures are unrelated. Other lanes e.g. serial-cpu-manager exercise this flow correctly.

**Comment by SergeyKanzhelev** [comment]
> @SergeyKanzhelev ready for review!

Is it ready now? Can you please squash commits

**Comment by ffromani** [comment]
> > @SergeyKanzhelev ready for review!
> 
> Is it ready now? Can you please squash commits

yes, but squashing commits anyway

**Comment by ffromani** [comment]
@SergeyKanzhelev squashed!

**Comment by ffromani** [comment]
after the squash, https://github.com/kubernetes/kubernetes/pull/127506 needs to be rebased on top of this one

**Comment by ffromani** [comment]
gentle ping because we have quite some PRs depending on this one

File: pkg/kubelet/cm/topologymanager/policy.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/cm/topologymanager/policy.go:36]: I have hard time judging if this TODO is fine to keep. If we do not plan to fix it any time soon, perhaps it can be rephrased into the explanation of the edge cases?

File: pkg/kubelet/metrics/metrics.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/metrics/metrics.go:144]: do we refer to numa node as zone in other places? Is zone the same as node?

File: pkg/kubelet/metrics/metrics.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/metrics/metrics.go:932]: Can you put it next to `CPUManagerPinningRequestsTotal` metric? They seems to be related. Also what is the difference between `NewCounterVec` and `NewCounter`?

File: pkg/kubelet/metrics/metrics.go
Comment (SergeyKanzhelev) [inline pkg/kubelet/metrics/metrics.go:936]: are we accounting for a failed admissions somehow? Or only for successful?

File: test/e2e_node/cpu_manager_metrics_test.go
Comment (SergeyKanzhelev) [inline test/e2e_node/cpu_manager_metrics_test.go:99]: How do you plan to use the `[pinning]` tag?

File: pkg/kubelet/cm/topologymanager/policy.go
Comment (ffromani) [reply pkg/kubelet/cm/topologymanager/policy.go:36]: good point, rephrasing

File: pkg/kubelet/metrics/metrics.go
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:144]: AFAIK "numa zones" "numa cells" and "numa nodes" are all synonyms, and I'd like to avoid the overinflaction of the "node" term.

File: pkg/kubelet/metrics/metrics.go
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:932]: sure, moving.
Regarding Counter vs CounterVec:
CounterVec bundles a set of Counters that all share the same Description. We use it because we want to count the same value partitioned by various dimensions: scope and boundary.

File: pkg/kubelet/metrics/metrics.go
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:936]: yes, using `CPUManagerPinningErrorsTotal` and `TopologyManagerAdmissionErrorsTotal`

File: test/e2e_node/cpu_manager_metrics_test.go
Comment (ffromani) [reply test/e2e_node/cpu_manager_metrics_test.go:99]: It's the easiest and simplest way I know to run a subset of tests without using crazy regexes using `gingko --focus`

File: pkg/kubelet/metrics/metrics.go
Comment (SergeyKanzhelev) [reply pkg/kubelet/metrics/metrics.go:144]: Do we ever use NUMA zone term? I see only numa nodes for example here: https://kubernetes.io/docs/tasks/administer-cluster/memory-manager/ Will it be more confusing to mix terms interchangeably than we gain by not mixing "nodes" and 
"NUMA nodes"?

File: test/e2e_node/cpu_manager_metrics_test.go
Comment (SergeyKanzhelev) [reply test/e2e_node/cpu_manager_metrics_test.go:99]: My question is whether you want to use it just locally or also in test-infra? We prefer to switch to the ginkgo tags, and we started moving there. Ginkgo tags make it easier to query which tests to run

Comment (SergeyKanzhelev) [reply test/e2e_node/cpu_manager_metrics_test.go:99]: If it is just for local debugging - I would question the value of it being added permanently. Is it easy to understand what pinning vs. alignment means? I have difficulty understanding it

File: pkg/kubelet/metrics/metrics.go
Comment (ffromani) [reply pkg/kubelet/metrics/metrics.go:144]: ok, changing to "NUMA nodes"

File: test/e2e_node/cpu_manager_metrics_test.go
Comment (ffromani) [reply test/e2e_node/cpu_manager_metrics_test.go:99]: I can move to labels if it helps

Comment (ffromani) [reply test/e2e_node/cpu_manager_metrics_test.go]: actually removed

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
